File: sikkerhet/src/test/java/no/nav/melosys/sikkerhet/abac/PepTest.java
package no.nav.melosys.sikkerhet.abac;

import no.nav.freg.abac.core.annotation.context.AbacContext;
import no.nav.freg.abac.core.dto.request.XacmlRequest;
import no.nav.freg.abac.core.dto.response.Decision;
import no.nav.freg.abac.core.dto.response.XacmlResponse;
import no.nav.freg.abac.core.service.AbacService;
import no.nav.melosys.exception.SikkerhetsbegrensningException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class PepTest {

    @InjectMocks
    public PepImpl pep;

    @Mock
    private AbacService abacService;

    @Mock
    private AbacContext abacContext;

    @Mock
    private XacmlResponse abacResponse;


    @BeforeEach
    public void setUp() {
        when(abacContext.getRequest()).thenReturn(new XacmlRequest());
        when(abacService.evaluate(any())).thenReturn(abacResponse);
    }

    @Test
    void testSjekkTilgangTilFnr() {
        when(abacResponse.getDecision()).thenReturn(Decision.PERMIT);
        pep.sjekkTilgangTilFnr("12345678910");
    }

    @Test
    void testSjekkTilgangTilFnrResponsDeny() {
        when(abacResponse.getDecision()).thenReturn(Decision.DENY);
        assertThatExceptionOfType(SikkerhetsbegrensningException.class)
            .isThrownBy(() -> pep.sjekkTilgangTilFnr("12345678910"))
            .withMessageContaining("ikke tilgang");
    }

    @Test
    void testSjekkTilgangTilFnrResponsIndeterminate() {
        when(abacResponse.getDecision()).thenReturn(Decision.INDETERMINATE);
        assertThatExceptionOfType(SikkerhetsbegrensningException.class)
            .isThrownBy(() -> pep.sjekkTilgangTilFnr("12345678910"))
            .withMessageContaining("ikke tilgang");
    }

    @Test
    void testSjekkTilgangTilFnrResponsNotApplicable() {
        when(abacResponse.getDecision()).thenReturn(Decision.NOT_APPLICABLE);
        assertThatExceptionOfType(SikkerhetsbegrensningException.class)
            .isThrownBy(() -> pep.sjekkTilgangTilFnr("12345678910"))
            .withMessageContaining("ikke tilgang");
    }


    @Test
    void testSjekkTilgangTilAktor() {
        when(abacResponse.getDecision()).thenReturn(Decision.PERMIT);
        pep.sjekkTilgangTilAktoerId("12345678910");
    }

    @Test
    void testSjekkTilgangTilAktorIdResponseDeny() {
        when(abacResponse.getDecision()).thenReturn(Decision.DENY);
        assertThatExceptionOfType(SikkerhetsbegrensningException.class)
            .isThrownBy(() -> pep.sjekkTilgangTilFnr("12345678910"))
            .withMessageContaining("ikke tilgang");
    }

    @Test
    void testSjekkTilgangTilAktorIdResponseIndeterminate() {
        when(abacResponse.getDecision()).thenReturn(Decision.INDETERMINATE);
        assertThatExceptionOfType(SikkerhetsbegrensningException.class)
            .isThrownBy(() -> pep.sjekkTilgangTilFnr("12345678910"))
            .withMessageContaining("ikke tilgang");
    }

    @Test
    void testSjekkTilgangTilAktorIdResponseNotApplicable() {
        when(abacResponse.getDecision()).thenReturn(Decision.NOT_APPLICABLE);
        assertThatExceptionOfType(SikkerhetsbegrensningException.class)
            .isThrownBy(() -> pep.sjekkTilgangTilFnr("12345678910"))
            .withMessageContaining("ikke tilgang");
    }
}

File: sikkerhet/src/test/kotlin/no/nav/melosys/sikkerhet/abac/PepTest.kt
package no.nav.melosys.sikkerhet.abac

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.string.shouldContain
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import no.nav.freg.abac.core.annotation.context.AbacContext
import no.nav.freg.abac.core.dto.request.XacmlRequest
import no.nav.freg.abac.core.dto.response.Decision
import no.nav.freg.abac.core.dto.response.XacmlResponse
import no.nav.freg.abac.core.service.AbacService
import no.nav.melosys.exception.SikkerhetsbegrensningException
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
class PepTest {

    @InjectMockKs
    lateinit var pep: PepImpl

    @MockK
    private lateinit var abacService: AbacService

    @MockK
    private lateinit var abacContext: AbacContext

    @MockK(relaxed = true)
    private lateinit var abacResponse: XacmlResponse

    @BeforeEach
    fun setUp() {
        every { abacContext.request } returns XacmlRequest()
        every { abacService.evaluate(any()) } returns abacResponse
    }

    @Test
    fun `sjekk tilgang til fnr med PERMIT skal ikke kaste exception`() {
        every { abacResponse.decision } returns Decision.PERMIT
        pep.sjekkTilgangTilFnr("12345678910")
    }

    @Test
    fun `sjekk tilgang til fnr med DENY skal kaste SikkerhetsbegrensningException`() {
        every { abacResponse.decision } returns Decision.DENY
        val exception = shouldThrow<SikkerhetsbegrensningException> {
            pep.sjekkTilgangTilFnr("12345678910")
        }
        exception.message shouldContain "ikke tilgang"
    }

    @Test
    fun `sjekk tilgang til fnr med INDETERMINATE skal kaste SikkerhetsbegrensningException`() {
        every { abacResponse.decision } returns Decision.INDETERMINATE
        val exception = shouldThrow<SikkerhetsbegrensningException> {
            pep.sjekkTilgangTilFnr("12345678910")
        }
        exception.message shouldContain "ikke tilgang"
    }

    @Test
    fun `sjekk tilgang til fnr med NOT_APPLICABLE skal kaste SikkerhetsbegrensningException`() {
        every { abacResponse.decision } returns Decision.NOT_APPLICABLE
        val exception = shouldThrow<SikkerhetsbegrensningException> {
            pep.sjekkTilgangTilFnr("12345678910")
        }
        exception.message shouldContain "ikke tilgang"
    }

    @Test
    fun `sjekk tilgang til aktor med PERMIT skal ikke kaste exception`() {
        every { abacResponse.decision } returns Decision.PERMIT
        pep.sjekkTilgangTilAktoerId("12345678910")
    }

    @Test
    fun `sjekk tilgang til aktor med DENY skal kaste SikkerhetsbegrensningException`() {
        every { abacResponse.decision } returns Decision.DENY
        val exception = shouldThrow<SikkerhetsbegrensningException> {
            pep.sjekkTilgangTilFnr("12345678910")
        }
        exception.message shouldContain "ikke tilgang"
    }

    @Test
    fun `sjekk tilgang til aktor med INDETERMINATE skal kaste SikkerhetsbegrensningException`() {
        every { abacResponse.decision } returns Decision.INDETERMINATE
        val exception = shouldThrow<SikkerhetsbegrensningException> {
            pep.sjekkTilgangTilFnr("12345678910")
        }
        exception.message shouldContain "ikke tilgang"
    }

    @Test
    fun `sjekk tilgang til aktor med NOT_APPLICABLE skal kaste SikkerhetsbegrensningException`() {
        every { abacResponse.decision } returns Decision.NOT_APPLICABLE
        val exception = shouldThrow<SikkerhetsbegrensningException> {
            pep.sjekkTilgangTilFnr("12345678910")
        }
        exception.message shouldContain "ikke tilgang"
    }
}
